code/noRAM/app/src/main/java/com/example/noram/PushNotificationService.java

```diff
@@ -0,0 +1,156 @@
+
+import android.app.AlertDialog;
+import android.util.Log;
+import android.widget.Toast;
+
+import static android.content.ContentValues.TAG;
+
+import androidx.annotation.NonNull;
+
+import com.example.noram.model.Event;
+import com.google.firebase.messaging.FirebaseMessagingService;
+import com.google.firebase.messaging.RemoteMessage;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+/**
+ * The PushNotificationService class is used to send notifications to attendees of an event.
+ * A {@link FirebaseMessagingService} subclass.
+ * @Maintainer Christiaan
+ * @Author Christiaan
+ */
+public class PushNotificationService extends FirebaseMessagingService {
+
+    /**
+     * Called when a new token is generated. Will update the FCM token of the attendee.
+     * @param token the new token
+     */
+    @Override
+    public void onNewToken(@NonNull String token) {
+        Log.d(TAG, "Refreshed token: " + token);
+        MainActivity.attendee.setFCMToken(token);
+    }
+
+    /**
+     * Called when a message is received. Will display a notification with the message received.
+     * @param remoteMessage the message received
+     */
+    @Override
+    public void onMessageReceived(RemoteMessage remoteMessage) {
+
+        Log.d(TAG, "From: " + remoteMessage.getFrom());
+        Log.d(TAG, "Message Notification Title: " + Objects.requireNonNull(remoteMessage.getNotification()).getTitle());
+        Log.d(TAG, "Message Notification Body: " + remoteMessage.getNotification().getBody());
+
+        // Change to the UI Thread
+        MainActivity.mn.runOnUiThread(() -> {
+            // Display the notification with an alert dialog
+            AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.mn);
+            builder.setTitle(remoteMessage.getNotification().getTitle());
+            builder.setMessage(remoteMessage.getNotification().getBody());
+            builder.setPositiveButton("OK", (dialog, which) -> dialog.dismiss());
+            builder.show();
+        });
+
+    }
+
+    /**
+     * A method to send a notification to the attendees of an event
+     * @param title the title of the notification
+     * @param data the data of the notification
+     * @param event the event to send the notification to
+     */
+    public void sendNotification(String title, String data, Event event) {
+
+        // Get the list of attendees
+
+        List<String> attendeeList = event.getCheckedInAttendees();
+        assert attendeeList != null;
+
+        // Send a notification to each attendee
+
+        for (String attendeeID : attendeeList) {
+
+            Log.d("AttendeeID", attendeeID);
+
+            MainActivity.db.getAttendeeRef().document(attendeeID).get().addOnCompleteListener(task -> {
```

Comment:
these could probably be onSuccessListeners since you don't do anything if they aren't successful - not sure if there was a reason you used onComplete?

Reply:
for some weird reason on success just did not work

Reply:
weird, but if thats the case then ok